### ISSUE TITLE
2opt deoptimizes from greedy solution?

### ISSUE THREAD

**Issue by sponsored** [estan/tsp#1]
Hi, not sure if this repo is being maintained, but just in case anyone else finds it on here and tries to use it. 

I haven't tried to track it down too much - I may be doing something stupid myself - but it seems that it is possible for the track to actually be deoptimized in the optimization process:

I've added some logging to debug the issue - running a simple test file sometimes generates an optimal tour, sometimes it doesn't. 

** nb. My comments inline marked by #<  **

```git diff .
diff --git a/tsp.cpp b/tsp.cpp
index 86a5fd5..0ce5b1f 100644
--- a/tsp.cpp
+++ b/tsp.cpp
@@ -9,11 +9,7 @@

 #include "matrix.h"

-#ifdef DEBUG
-    #define LOG(x) std::cerr << x << std::endl
-#else
-    #define LOG(x)
-#endif
+#define LOG(x) std::cerr << x << std::endl

 using namespace std;

@@ -234,6 +230,13 @@ inline vector<uint16_t> greedy(const Matrix<uint32_t>& d) {
         tour[i] = k;
         used[k] = true;
     }
+    std::cerr << "Greedy: tour = " << std::endl;
+    for(const auto t : tour)
+           std::cerr << " " << t << std::endl;
+
+    const int len = length(tour, d);
+    std::cerr << "Greedy length: " << len << std::endl;
+
     return tour;
 }

@@ -560,6 +563,7 @@ vector<uint16_t> approximate(istream &in, const chrono::time_point<T>& deadline)
     LOG("  iterations: " << i);
     LOG("  totalTime: " << totalTime);
     LOG("  averageTime: " << averageTime);
+    LOG("  length: " << shortestTourLength);

     return shortestTour;
 }



~/tsp$ t test2.tsp
9
90 90
90 80
90 70
120 110
120 120
120 130
130 110
130 120
130 130



~/tsp$ cat test2.tsp | ./tsp
Greedy: tour =
 0
 1
 2
 3
 4
 5
 8
 7
 6
Greedy length: 165   #< length is also inflated both in this any my previous testing, not sure why
Main Loop Statistics
  iterations: 839488
  totalTime: 0 ms
  averageTime: 0 ms
  length: 162            #< In other testing this is less even when the optimization produces the same tour
0
4                              #< bypasses 3!
5
8
7
6
3
2
1```

**Comment by estan** [comment]
@sponsored Wow I had forgot I had this old school assignment solution posted on GitHub. Thanks for filing this issue :)

Applying 2-opt followed by 3-opt to a tour is not guaranteed to make it shorter. That's why the code only updates `shortestTour` and `shortestTourLength` if applying 2-opt + 3-opt actually improved the tour:

```cpp
        // Optimize tour with 2-opt + 3-opt.
        twoOpt(tour, d, neighbor, position, max, min);
        threeOpt(tour, d, neighbor, position, max, min, threeOptDeadline);

        uint64_t tourLength = length(tour, d);
        if (tourLength < shortestTourLength) {
            // Shorter tour found.
            shortestTour = tour;
            shortestTourLength = tourLength;
        }
```

Regarding your comment

    Greedy length: 165   #< length is also inflated both in this any my previous testing, not sure why

Can you explain what you mean by inflated? The printout you added is the length of the initial greedy tour 0 1 2 3 4 5 8 7 6.

We can see from the distance matrix that this is the greedy solution

| City | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | Greedy Tour |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 0 | 0 | 10 | 20 | 36 | 42 | 50 | 45 | 50 | 57 | Step 1: Go to city 1 since 10 is shortest |
| 1 | 10 | 0 | 10 | 42 | 50 | 58 | 50 | 57 | 64 | Step 2: Go to city 2 since 10 is shortest |
| 2 | 20 | 10 | 0 | 50 | 58 | 67 | 57 | 64 | 72 | Step 3: Go to city 3 since 50 is shortest |
| 3 | 36 | 42 | 50 | 0 | 10 | 20 | 10 | 14 | 22 | Step 4: Go to city 4 since 10 is shortest |
| 4 | 42 | 50 | 58 | 10 | 0 | 10 | 14 | 10 | 14 | Step 5: Go to city 5 since 10 is shortest |
| 5 | 50 | 58 | 67 | 20 | 10 | 0 | 22 | 14 | 10 | Step 6: Go to city 8 since 10 is shortest |
| 6 | 45 | 50 | 57 | 10 | 14 | 22 | 0 | 10 | 20 | Step 9: Go to city 0 since 45 is shortest |
| 7 | 50 | 57 | 64 | 14 | 10 | 14 | 10 | 0 | 10 | Step 8: Go to city 6 since 10 is shortest |
| 8 | 57 | 64 | 72 | 22 | 14 | 10 | 20 | 10 | 0 | Step 7: Go to city 7 since 10 is shortest |

Summing up the distances we get 10 + 10 + 50 + 10 + 10 + 10 + 45 + 10 + 10 = 165, so 165 is the length of the greedy solution.

One thing to mention is that, as pointed out in the README.md, this is a solution to the "Traveling Salesperson 2D" problem on the Kattis problem server at my school at the time. One special thing about that problem is that the distances between cities is defined as the Euclidian distance *rounded to the nearest integer*.

I realize that since not everyone has access to that problem server, I should probably post the full problem in the README.md. Anyway it seems I still have access even 8 years after graduating, so here is the full problem statement:

![tsp2d](https://github.com/estan/tsp/assets/86675/eaed8495-e9c8-452e-941e-6f704683685f)

**Comment by estan** [comment]
I've added the full problem statement to README.md now.

**Comment by estan** [comment]
@sponsored Actually, you did find an issue, see https://github.com/estan/tsp/pull/2

**Comment by sponsored** [comment]
Hi, thanks, and yes sorry that's incorrect about the length providing an incorrect readout. In my testing my own distance calcs differ but that may be issues in my code.
